src/node/plugins/offlineSearchPlugin.ts

```diff
@@ -0,0 +1,272 @@
+import path from 'node:path'
+import type { Plugin, ViteDevServer } from 'vite'
+import MiniSearch from 'minisearch'
+import fs from 'fs-extra'
+import _debug from 'debug'
+import type { SiteConfig } from '../config'
+import { createMarkdownRenderer } from '../markdown/markdown.js'
+import { resolveSiteDataByRoute } from '../shared.js'
+
+const debug = _debug('vitepress:offline-search')
+
+const OFFLINE_SEARCH_INDEX_ID = '@offlineSearchIndex'
+const OFFLINE_SEARCH_INDEX_REQUEST_PATH = '/' + OFFLINE_SEARCH_INDEX_ID
+
+interface IndexObject {
+  id: string
+  text: string
+  title: string
+  titles: string[]
+}
+
+export async function offlineSearchPlugin(
+  siteConfig: SiteConfig
+): Promise<Plugin> {
+  if (
+    siteConfig.userConfig.themeConfig?.algolia ||
+    !siteConfig.userConfig.themeConfig?.offlineSearch
+  ) {
+    return {
+      name: 'vitepress:offline-search',
+      resolveId(id) {
+        if (id.startsWith(OFFLINE_SEARCH_INDEX_ID)) {
+          return `/${id}`
+        }
+      },
+      load(id) {
+        if (id.startsWith(OFFLINE_SEARCH_INDEX_REQUEST_PATH)) {
+          return `export default '{}'`
+        }
+      }
+    }
+  }
+
+  const md = await createMarkdownRenderer(
+    siteConfig.srcDir,
+    siteConfig.userConfig.markdown,
+    siteConfig.userConfig.base,
+    siteConfig.logger
+  )
+
+  const indexByLocales = new Map<string, MiniSearch<IndexObject>>()
+
+  function getIndexByLocale(locale: string) {
+    let index = indexByLocales.get(locale)
+    if (!index) {
+      index = new MiniSearch<IndexObject>({
+        fields: ['title', 'titles', 'text'],
+        storeFields: ['title', 'titles']
+      })
+      indexByLocales.set(locale, index)
+    }
+    return index
+  }
+
+  function getLocaleForPath(file: string) {
+    const relativePath = path.relative(siteConfig.srcDir, file)
+    const siteData = resolveSiteDataByRoute(siteConfig.site, relativePath)
+    return siteData?.localeIndex ?? 'root'
+  }
+
+  function getIndexForPath(file: string) {
+    const locale = getLocaleForPath(file)
+    return getIndexByLocale(locale)
+  }
+
+  let server: ViteDevServer | undefined
+
+  async function onIndexUpdated() {
+    if (server) {
+      server.moduleGraph.onFileChange(OFFLINE_SEARCH_INDEX_REQUEST_PATH)
+      // HMR
+      const mod = server.moduleGraph.getModuleById(
+        OFFLINE_SEARCH_INDEX_REQUEST_PATH
+      )
+      if (!mod) return
+      server.ws.send({
+        type: 'update',
+        updates: [
+          {
+            acceptedPath: mod.url,
+            path: mod.url,
+            timestamp: Date.now(),
+            type: 'js-update'
+          }
+        ]
+      })
+    }
+  }
+
+  function getDocId(file: string) {
+    let id = file.replace(siteConfig.srcDir, siteConfig.userConfig.base ?? '')
+    id = id.replace(/\.md$/, siteConfig.cleanUrls ? '' : '.html')
+    return id
```

Comment:
While testing I figured 2 small issues.  The plugin 
- doesn't work w/ [route rewrites](https://vitepress.dev/guide/routing#route-rewrites)
- produces a double `//` in the `id`

The code below fixes both issues

```suggestion
    let relFile = path.relative(siteConfig.srcDir, file)
    relFile = siteConfig.rewrites.map[relFile] || relFile
    let id = path.join(siteConfig.userConfig.base ?? "", relFile);
    id = id.replace(/\.md$/, siteConfig.cleanUrls ? "" : ".html");
    return id;
```